File: app/src/main/java/com/adam/apkmaker/MainActivity.java
package com.adam.apkmaker;

import android.app.*;
import android.os.*;
import android.content.*;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.net.Uri;
import android.provider.Settings;
import android.view.*;
import android.view.inputmethod.InputMethodManager;
import android.widget.*;
import java.util.concurrent.TimeUnit;

public class MainActivity extends Activity {
    private LinearLayout root, content;
    private TextView fileStatus, buildStatus, timerText;
    private String selectedFile = "";
    private boolean fastActive = false;
    private long fastUntil = 0L;
    private static final int PICK_FILE = 501;
    private static final int PICK_ICON = 502;

    private int dp(float v) { return (int)(v * getResources().getDisplayMetrics().density + 0.5f); }

    private TextView tv(String text, float size, int color) {
        TextView t = new TextView(this);
        t.setText(text); t.setTextSize(size); t.setTextColor(color);
        t.setGravity(Gravity.CENTER_VERTICAL);
        return t;
    }

    private GradientDrawable bg(int color, int radius, int strokeColor) {
        GradientDrawable g = new GradientDrawable();
        g.setColor(color); g.setCornerRadius(dp(radius));
        if (strokeColor != 0) g.setStroke(dp(1), strokeColor);
        return g;
    }

    private Button button(String text, boolean teal) {
        Button b = new Button(this);
        b.setText(text); b.setTextSize(15); b.setTextColor(Color.rgb(244,247,248));
        b.setAllCaps(false); b.setPadding(dp(10),0,dp(10),0);
        b.setBackground(bg(teal ? Color.rgb(23,58,55) : Color.rgb(29,33,37), 18,
                teal ? Color.rgb(86,230,209) : Color.rgb(41,50,56)));
        return b;
    }

    private LinearLayout card() {
        LinearLayout c = new LinearLayout(this);
        c.setOrientation(LinearLayout.VERTICAL); c.setPadding(dp(18),dp(16),dp(18),dp(16));
        c.setBackground(bg(Color.rgb(21,24,27),24,Color.rgb(36,41,46)));
        LinearLayout.LayoutParams lp = new LinearLayout.LayoutParams(-1,-2);
        lp.setMargins(0,dp(10),0,dp(10)); c.setLayoutParams(lp); return c;
    }

    @Override public void onCreate(Bundle b) {
        super.onCreate(b);
        getWindow().setStatusBarColor(Color.rgb(9,11,13));
        getWindow().setNavigationBarColor(Color.rgb(9,11,13));
        buildUi();
    }

    private void buildUi() {
        root = new LinearLayout(this); root.setOrientation(LinearLayout.VERTICAL);
        root.setBackgroundColor(Color.rgb(9,11,13));
        root.setPadding(dp(18),dp(10),dp(18),dp(8));

        LinearLayout top = new LinearLayout(this); top.setGravity(Gravity.CENTER_VERTICAL);
        TextView brand = tv("ADAM APK MAKER",22,Color.rgb(244,247,248));
        brand.setTypeface(null,1); top.addView(brand,new LinearLayout.LayoutParams(0,dp(60),1));

        Button help = button("?",false);
        Button key = button("KEY",false);
        Button ai = button("AI",false);
        top.addView(key,new LinearLayout.LayoutParams(dp(58),dp(48)));
        top.addView(help,new LinearLayout.LayoutParams(dp(48),dp(48)));
        top.addView(ai,new LinearLayout.LayoutParams(dp(48),dp(48)));
        root.addView(top);

        TextView sub=tv("ZIP / HTML  •  Android APK Builder  •  Made by ADAM",12,Color.rgb(137,146,152));
        root.addView(sub,new LinearLayout.LayoutParams(-1,dp(28)));

        LinearLayout tabs = new LinearLayout(this); tabs.setPadding(0,dp(4),0,dp(4));
        Button build=button("Build",true), options=button("Options",false), history=button("History",false);
        tabs.addView(build,new LinearLayout.LayoutParams(0,dp(48),1));
        tabs.addView(options,new LinearLayout.LayoutParams(0,dp(48),1));
        tabs.addView(history,new LinearLayout.LayoutParams(0,dp(48),1));
        root.addView(tabs);

        ScrollView scroll=new ScrollView(this);
        content=new LinearLayout(this); content.setOrientation(LinearLayout.VERTICAL);
        scroll.addView(content,new ScrollView.LayoutParams(-1,-2)); root.addView(scroll,new LinearLayout.LayoutParams(-1,0,1));

        fileStatus=tv("No file selected",14,Color.rgb(137,146,152));
        LinearLayout upload=card();
        TextView h=tv("PROJECT SOURCE",12,Color.rgb(86,230,209)); h.setTypeface(null,1); upload.addView(h);
        Button pick=button("Select ZIP or HTML file",true); upload.addView(pick,new LinearLayout.LayoutParams(-1,dp(58)));
        upload.addView(fileStatus,new LinearLayout.LayoutParams(-1,dp(38)));
        TextView hint=tv("ZIP projects and standalone HTML files are supported.",12,Color.rgb(137,146,152));
        upload.addView(hint);
        content.addView(upload);

        LinearLayout iconCard=card();
        TextView ih=tv("APP ICON",12,Color.rgb(86,230,209)); ih.setTypeface(null,1); iconCard.addView(ih);
        Button icon=button("Add App Icon",false); iconCard.addView(icon,new LinearLayout.LayoutParams(-1,dp(58)));
        TextView it=tv("PNG, JPG or WEBP • square image recommended",12,Color.rgb(137,146,152)); iconCard.addView(it);
        content.addView(iconCard);

        LinearLayout details=card();
        TextView dh=tv("APP DETAILS",12,Color.rgb(86,230,209)); dh.setTypeface(null,1); details.addView(dh);
        EditText name=new EditText(this); name.setHint("App name"); name.setTextColor(Color.WHITE); name.setHintTextColor(Color.rgb(100,110,116)); name.setSingleLine(); name.setBackground(bg(Color.rgb(15,18,20),18,Color.rgb(58,76,80))); name.setPadding(dp(16),0,dp(16),0);
        details.addView(name,new LinearLayout.LayoutParams(-1,dp(56)));
        EditText pkg=new EditText(this); pkg.setHint("Package name  •  com.example.app"); pkg.setTextColor(Color.WHITE); pkg.setHintTextColor(Color.rgb(100,110,116)); pkg.setSingleLine(); pkg.setBackground(bg(Color.rgb(15,18,20),18,Color.rgb(58,76,80))); pkg.setPadding(dp(16),0,dp(16),0);
        LinearLayout.LayoutParams plp=new LinearLayout.LayoutParams(-1,dp(56)); plp.setMargins(0,dp(10),0,0); details.addView(pkg,plp);
        content.addView(details);

        LinearLayout fast=card();
        LinearLayout row=new LinearLayout(this); row.setGravity(Gravity.CENTER_VERTICAL);
        TextView ft=tv("Fast Build",17,Color.WHITE); ft.setTypeface(null,1);
        timerText=tv("Key required",12,Color.rgb(137,146,152)); timerText.setGravity(Gravity.RIGHT|Gravity.CENTER_VERTICAL);
        row.addView(ft,new LinearLayout.LayoutParams(0,dp(48),1)); row.addView(timerText,new LinearLayout.LayoutParams(dp(110),dp(48)));
        fast.addView(row);
        Button buildFast=button("⚡  Fast Build",false); fast.addView(buildFast,new LinearLayout.LayoutParams(-1,dp(54)));
        content.addView(fast);

        LinearLayout bottom=new LinearLayout(this); bottom.setPadding(0,dp(8),0,0);
        Button buildApk=button("Build APK",true); bottom.addView(buildApk,new LinearLayout.LayoutParams(-1,dp(60)));
        root.addView(bottom);

        pick.setOnClickListener(v -> chooseProject());
        icon.setOnClickListener(v -> chooseIcon());
        key.setOnClickListener(v -> showKeyDialog());
        ai.setOnClickListener(v -> showAiDialog());
        help.setOnClickListener(v -> showHelp());
        build.setOnClickListener(v -> runBuild(false));
        options.setOnClickListener(v -> showOptions());
        history.setOnClickListener(v -> showHistory());
        buildApk.setOnClickListener(v -> runBuild(false));
        buildFast.setOnClickListener(v -> runBuild(true));

        setContentView(root);
        startTimerLoop();
    }

    private void chooseProject() {
        Intent i=new Intent(Intent.ACTION_OPEN_DOCUMENT); i.addCategory(Intent.CATEGORY_OPENABLE); i.setType("*/*");
        i.putExtra(Intent.EXTRA_MIME_TYPES,new String[]{"application/zip","text/html","text/plain","application/octet-stream"});
        startActivityForResult(i,PICK_FILE);
    }
    private void chooseIcon() {
        Intent i=new Intent(Intent.ACTION_OPEN_DOCUMENT); i.addCategory(Intent.CATEGORY_OPENABLE); i.setType("image/*"); startActivityForResult(i,PICK_ICON);
    }
    @Override protected void onActivityResult(int r,int c,Intent d) {
        super.onActivityResult(r,c,d); if(c!=RESULT_OK||d==null)return;
        Uri u=d.getData(); if(u==null)return;
        if(r==PICK_FILE){ selectedFile=u.toString(); fileStatus.setText("Selected: "+u.getLastPathSegment()); fileStatus.setTextColor(Color.rgb(86,230,209)); }
        else if(r==PICK_ICON) Toast.makeText(this,"App icon selected",Toast.LENGTH_SHORT).show();
    }

    private void showKeyDialog() {
        final EditText input=new EditText(this); input.setSingleLine(); input.setHint("Enter access key");
        input.setTextColor(Color.WHITE); input.setHintTextColor(Color.GRAY);
        input.setInputType(1);
        LinearLayout box=new LinearLayout(this); box.setPadding(dp(24),0,dp(24),0); box.addView(input,new LinearLayout.LayoutParams(-1,dp(56)));
        AlertDialog dialog=new AlertDialog.Builder(this).setTitle("Fast Build Key").setMessage("Enter your access key to unlock Fast Build for 5 minutes.").setView(box).setNegativeButton("Cancel",null).setPositiveButton("Activate",null).create();
        dialog.setOnShowListener(x -> dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener(v -> {
            if("ADAM31".equals(input.getText().toString().trim())){
                fastActive=true; fastUntil=System.currentTimeMillis()+TimeUnit.MINUTES.toMillis(5);
                timerText.setText("05:00 active"); dialog.dismiss();
                Toast.makeText(this,"Fast Build activated for 5 minutes",Toast.LENGTH_SHORT).show();
            } else input.setError("Invalid key");
        }));
        dialog.show();
    }

    private void showAiDialog() {
        TextView t=tv("AI Assistant\n\nI can help analyze your selected ZIP/HTML project, suggest app settings, and explain build errors.\n\nCloud AI can be connected later with your own API key.",15,Color.WHITE);
        t.setPadding(dp(24),dp(10),dp(24),dp(10));
        new AlertDialog.Builder(this).setTitle("AI Assistant").setView(t).setPositiveButton("Close",null).show();
    }
    private void showHelp() {
        new AlertDialog.Builder(this).setTitle("How it works").setMessage("1. Select a ZIP or HTML file.\n2. Add an icon and app details.\n3. Choose Build APK or Fast Build.\n4. The builder service packages the project and returns an APK.\n\nNote: actual APK compilation requires a connected build service or a full Android build toolchain; this app does not fake an APK.").setPositiveButton("OK",null).show();
    }
    private void showOptions() {
        new AlertDialog.Builder(this).setTitle("Options").setItems(new String[]{"Internet permission: ON","Portrait orientation","Release signing","Offline cache: ON"} ,null).setPositiveButton("Done",null).show();
    }
    private void showHistory() {
        new AlertDialog.Builder(this).setTitle("Build History").setMessage("No completed builds yet.").setPositiveButton("OK",null).show();
    }

    private void runBuild(boolean fast) {
        if(selectedFile.isEmpty()){ Toast.makeText(this,"Select a ZIP or HTML file first",Toast.LENGTH_SHORT).show(); return; }
        if(fast && !fastActive){ Toast.makeText(this,"Enter the access key first",Toast.LENGTH_SHORT).show(); return; }
        int seconds=fast?60:300;
        final Dialog dialog=new Dialog(this);
        LinearLayout box=new LinearLayout(this); box.setOrientation(LinearLayout.VERTICAL); box.setPadding(dp(24),dp(22),dp(24),dp(18));
        TextView title=tv(fast?"Fast Build":"Build APK",20,Color.WHITE); title.setTypeface(null,1); box.addView(title);
        TextView stage=tv("Preparing…",14,Color.rgb(137,146,152)); box.addView(stage,new LinearLayout.LayoutParams(-1,dp(42)));
        ProgressBar bar=new ProgressBar(this,null,android.R.attr.progressBarStyleHorizontal); bar.setMax(100); box.addView(bar,new LinearLayout.LayoutParams(-1,dp(18)));
        TextView time=tv("05:00",12,Color.rgb(86,230,209)); time.setGravity(Gravity.CENTER); box.addView(time,new LinearLayout.LayoutParams(-1,dp(34)));
        dialog.setContentView(box); Window w=dialog.getWindow(); if(w!=null){w.setBackgroundDrawable(bg(Color.rgb(21,24,27),24,0));w.setLayout(-1,-2);}
        dialog.show();
        final String[] stages={"Preparing…","Analyzing project…","Building Android resources…","Compiling…","Packaging APK…","Signing…","Ready"};
        final long start=System.currentTimeMillis();
        final Handler h=new Handler(Looper.getMainLooper());
        Runnable r=new Runnable(){ public void run(){
            long elapsed=(System.currentTimeMillis()-start)/1000; int pct=(int)Math.min(100,elapsed*100/seconds);
            bar.setProgress(pct); int remain=Math.max(0,seconds-(int)elapsed);
            time.setText(String.format("%02d:%02d",remain/60,remain%60));
            stage.setText(stages[Math.min(stages.length-1,(int)(pct/17))]);
            if(elapsed<seconds){h.postDelayed(this,250);}
            else { stage.setText("Ready"); time.setText("100%"); Toast.makeText(MainActivity.this,"Build simulation complete. Connect a build service to produce a real APK.",Toast.LENGTH_LONG).show(); }
        }}; h.post(r);
    }

    private void startTimerLoop(){
        final Handler h=new Handler(Looper.getMainLooper());
        h.post(new Runnable(){public void run(){
            if(fastActive){
                long left=fastUntil-System.currentTimeMillis();
                if(left<=0){fastActive=false;timerText.setText("Key required");}
                else {long s=left/1000;timerText.setText(String.format("%02d:%02d active",s/60,s%60));}
            }
            h.postDelayed(this,1000);
        }});
    }
}

File: app/src/main/kotlin/com/tycept/wzh2apk/MainActivity.kt

package com.tycept.wzh2apk

import android.app.*
import android.content.*
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.net.Uri
import android.os.*
import android.view.*
import android.widget.*
import org.json.*
import java.io.*
import java.net.*
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.*
import java.util.zip.*

class MainActivity : Activity() {
    companion object { const val API="https://zip-to-apk-ce53.onrender.com"; const val PICK_PROJECT=100 }
    private val BG=Color.rgb(13,15,17); private val CARD=Color.rgb(21,23,26); private val PANEL=Color.rgb(27,29,33)
    private val MUTED=Color.rgb(157,161,168); private val WHITE=Color.rgb(238,240,242); private val ACCENT=Color.rgb(99,230,213)
    private lateinit var content: LinearLayout; private lateinit var footerLimit: TextView; private lateinit var buildButton: Button
    private var sourceMode="zip"; private var codeLang="java"; private var token:String?=null; private var jobId:String?=null; private var selectedUri:Uri?=null
    private var appName:EditText?=null; private var packageSuffix:EditText?=null; private var nativeCode:EditText?=null; private var htmlCode:EditText?=null; private var status:TextView?=null
    private val io=Executors.newCachedThreadPool(); private val handler=Handler(Looper.getMainLooper()); private val prefs by lazy { getSharedPreferences("wzh",0) }
    private val selectedPermissions=mutableListOf<String>()

    override fun onCreate(b:Bundle?){super.onCreate(b);window.statusBarColor=Color.rgb(12,14,16);shell();ensureSession();refreshLimit()}
    private fun tv(s:String,sp:Float,c:Int)=TextView(this).apply{ text=s;textSize=sp;setTextColor(c);setPadding(0,6,0,6)}
    private fun bg(c:Int,r:Float,stroke:Int=0,sc:Int=0)=GradientDrawable().apply{setColor(c);cornerRadius=r;if(stroke>0)setStroke(stroke,sc)}
    private fun btn(s:String)=Button(this).apply{text=s;textSize=13f;setTextColor(WHITE);isAllCaps=false;background=bg(PANEL,16f,1,Color.rgb(55,58,63));setPadding(12,4,12,4)}
    private fun edit(h:String)=EditText(this).apply{hint=h;setHintTextColor(Color.rgb(100,104,110));setTextColor(WHITE);textSize=15f;setSingleLine();setPadding(16,0,16,0);background=bg(Color.rgb(24,26,30),14f,1,Color.rgb(52,55,60))}
    private fun add(v:View,top:Int=0){content.addView(v,LinearLayout.LayoutParams(-1,-2).apply{setMargins(0,top,0,0)})}
    private fun shell(){
        val root=LinearLayout(this).apply{orientation=LinearLayout.VERTICAL;setBackgroundColor(BG)}
        val head=LinearLayout(this).apply{gravity=Gravity.CENTER_VERTICAL;setPadding(18,12,12,10);setBackgroundColor(Color.BLACK)}
        head.addView(tv("↗",28f,ACCENT).apply{gravity=Gravity.CENTER;background=bg(Color.BLACK,0f,2,WHITE)},LinearLayout.LayoutParams(58,58))
        head.addView(LinearLayout(this).apply{orientation=LinearLayout.VERTICAL;setPadding(16,0,0,0);addView(tv("WZH 2 APK",23f,WHITE));addView(tv(".zip  →  .apk  ·  by Tycept",13f,MUTED))},LinearLayout.LayoutParams(0,-2,1f))
        head.addView(btn("↻").apply{setTextSize(25f);setOnClickListener{ensureSession();refreshLimit()}},LinearLayout.LayoutParams(54,54));head.addView(btn("?").apply{setTextSize(20f);setOnClickListener{help()}},LinearLayout.LayoutParams(54,54));root.addView(head)
        val tabs=LinearLayout(this).apply{setPadding(8,8,8,8);setBackgroundColor(Color.rgb(17,19,22))};listOf("Build","Options","History","Progress").forEach{n->tabs.addView(btn(n).apply{setOnClickListener{show(n)}},LinearLayout.LayoutParams(0,48,1f))};root.addView(tabs)
        val scroll=ScrollView(this).apply{isFillViewport=true};content=LinearLayout(this).apply{orientation=LinearLayout.VERTICAL;setPadding(24,6,24,120)};scroll.addView(content);root.addView(scroll,LinearLayout.LayoutParams(-1,0,1f))
        val foot=LinearLayout(this).apply{orientation=LinearLayout.VERTICAL;setPadding(24,8,24,10);setBackgroundColor(Color.rgb(10,11,13))};footerLimit=tv("Builds today: —",13f,MUTED).apply{gravity=Gravity.CENTER};foot.addView(footerLimit);buildButton=btn("BUILD APK").apply{setTextSize(16f);isEnabled=false;setOnClickListener{startBuild()}};foot.addView(buildButton,LinearLayout.LayoutParams(-1,58));root.addView(foot);setContentView(root);show("Build")
    }
    private fun clear(){content.removeAllViews()}; private fun show(tab:String){clear();when(tab){"Build"->build();"Options"->options();"History"->history();else->progressTab()}}
    private fun section(label:String,desc:String){add(tv(label.uppercase(),12f,ACCENT),12);add(tv(desc,13f,MUTED),0)}
    private fun build(){add(tv("Build your app",27f,WHITE),12);add(tv("Turn a ZIP project, website, or Java/Kotlin source into an installable Android build.",14f,MUTED));section("Source","Choose what you're starting from.");listOf("Upload File" to "zip","Website link" to "url","Write code" to "code").forEach{(label,mode)->add(btn((if(sourceMode==mode)"✓  " else "")+label).apply{setOnClickListener{sourceMode=mode;show("Build")}},8)}
        when(sourceMode){"zip"->{add(btn("SELECT PROJECT ZIP / HTML").apply{setOnClickListener{pick()}},16);add(tv(selectedUri?.lastPathSegment?.let{"Selected: $it"} ?: "No project selected",14f,MUTED),4)}
            "url"->{val u=edit("https://example.com");u.setText(prefs.getString("remoteUrl",""));add(u,16);add(btn("USE THIS WEBSITE").apply{setOnClickListener{prefs.edit().putString("remoteUrl",u.text.toString().trim()).apply();buildButton.isEnabled=true}},8)}
            else->{val langs=LinearLayout(this);listOf("html","java","kotlin").forEach{l->langs.addView(btn((if(codeLang==l)"✓  " else "")+l.replaceFirstChar{it.uppercase()}).apply{setOnClickListener{codeLang=l;show("Build")}},LinearLayout.LayoutParams(0,52,1f))};add(langs,16);if(codeLang=="html"){htmlCode=code("<!doctype html>\n<html><body><h1>Hello from WZH</h1></body></html>");add(htmlCode!!,8)}else{nativeCode=code(if(codeLang=="java")javaTemplate(derivedPackage()) else kotlinTemplate(derivedPackage()));add(nativeCode!!,8)}add(btn("RESET TEMPLATE").apply{setOnClickListener{nativeCode?.setText(if(codeLang=="java")javaTemplate(derivedPackage()) else kotlinTemplate(derivedPackage()))}},8)}}
        buildButton.isEnabled=sourceMode=="url"||selectedUri!=null||(sourceMode=="code"&&((if(codeLang=="html")htmlCode else nativeCode)?.text?.isNotBlank()==true))
    }
    private fun code(s:String)=EditText(this).apply{text=s;textSize=13f;setTextColor(Color.rgb(220,225,228));typeface=Typeface.MONOSPACE;gravity=Gravity.TOP or Gravity.START;setPadding(14,14,14,14);background=bg(Color.rgb(10,12,14),14f,1,Color.rgb(54,58,63));minHeight=260;setSingleLine(false)}
    private fun derivedPackage():String{var s=(packageSuffix?.text?.toString()?.trim().takeUnless{it.isNullOrEmpty()} ?: appName?.text?.toString()?.trim().takeUnless{it.isNullOrEmpty()} ?: "myapp");s=s!!.lowercase(Locale.US).replace(Regex("[^a-z0-9]+"),".").replace(Regex("\\.+"),".");return "com."+s.trim('.')}
    private fun javaTemplate(p:String)="""package $p;

import android.app.Activity;
import android.os.Bundle;
import android.widget.TextView;

public class MainActivity extends Activity {
    @Override protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TextView tv = new TextView(this);
        tv.setText("Hello from Java!");
        tv.setTextSize(20);
        tv.setPadding(40,40,40,40);
        setContentView(tv);
    }
}
"""
    private fun kotlinTemplate(p:String)="""package $p

import android.app.Activity
import android.os.Bundle
import android.widget.TextView

class MainActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val tv = TextView(this)
        tv.text = "Hello from Kotlin!"
        tv.textSize = 20f
        tv.setPadding(40,40,40,40)
        setContentView(tv)
    }
}
"""
    private fun options(){add(tv("Options",27f,WHITE),12);add(tv("Configure the Android build without changing your project source.",14f,MUTED));section("App details","Identity used by the generated Android project.");appName=edit("App name").apply{setText(prefs.getString("appName",""))};packageSuffix=edit("Package suffix (optional)").apply{setText(prefs.getString("pkg",""))};add(appName!!,4);add(packageSuffix!!,8);add(edit("App link host (optional)"),8);add(edit("Google web client ID (optional)"),8);section("Protection","Build-time hardening options.");listOf("Harden with Dex2C","Sign release build","Fast cache").forEach{add(CheckBox(this).apply{text=it;setTextColor(WHITE);isChecked=it=="Sign release build"})};section("Startup loading","Select the splash animation and speed.");add(Spinner(this).apply{adapter=ArrayAdapter(this@MainActivity,android.R.layout.simple_spinner_dropdown_item,arrayOf("Off","Tumble","Fade","Typewriter","Pulse","Slide","Custom"))},4);add(Spinner(this).apply{adapter=ArrayAdapter(this@MainActivity,android.R.layout.simple_spinner_dropdown_item,arrayOf("Normal","Slow","Fast"))},4);section("Permissions","Optional Android permissions sent to the server.");val perms=arrayOf("Internet","Network state","Notifications","Keep screen awake","Background tasks","Biometric unlock","Camera","Microphone","Photos","Videos","Audio files","Precise location","Approximate location","Background location","Files & media (legacy)","Read contacts","Edit contacts","Read calendar","Edit calendar","Make calls","Phone state","Call log","Read SMS","Send SMS","Bluetooth","Nearby Wi-Fi devices","Body sensors","Physical activity","Vibrate");perms.forEach{p->add(CheckBox(this).apply{text=p;setTextColor(WHITE);isChecked=p in arrayOf("Internet","Network state","Camera","Microphone","Precise location","Approximate location","Files & media (legacy)","Vibrate")})};add(btn("SAVE OPTIONS").apply{setOnClickListener{prefs.edit().putString("appName",appName!!.text.toString()).putString("pkg",packageSuffix!!.text.toString()).apply();Toast.makeText(this,"Options saved",Toast.LENGTH_SHORT).show()}},20)}
    private fun history(){add(tv("Build history",27f,WHITE),12);add(tv("Your past builds from the WZH server.",14f,MUTED));val body=tv("Loading…",14f,MUTED);add(btn("REFRESH HISTORY").apply{setOnClickListener{loadHistory(body)}},14);add(body,12);loadHistory(body)}
    private fun progressTab(){add(tv("Build progress",27f,WHITE),12);status=tv(if(jobId==null)"No active build." else "Job: $jobId",15f,WHITE);add(status!!);add(tv("The server status is polled every two seconds.",14f,MUTED),8);add(btn("CANCEL BUILD").apply{setOnClickListener{cancelBuild()}},14)}

    private fun ensureSession(){io.execute{try{token=JSONObject(get("$API/api/session",null)).optString("token");ui("Build session ready")}catch(e:Exception){ui("Session: ${e.message}")}}}
    private fun refreshLimit(){io.execute{try{val o=JSONObject(get("$API/api/build-limit",null));runOnUiThread{footerLimit.text="Builds today: ${o.optInt("used")}/${o.optInt("limit")} used"}}catch(_:Exception){}}}
    private fun get(url:String,t:String?):String{val c=URL(url).openConnection() as HttpURLConnection;c.connectTimeout=30000;c.readTimeout=30000;if(t!=null)c.setRequestProperty("X-Session-Token",t);c.setRequestProperty("Accept","application/json");c.setRequestProperty("Origin",API);val code=c.responseCode;val s=read(if(code>=400)c.errorStream else c.inputStream);if(code !in 200..299)throw IOException("HTTP $code $s");return s}
    private fun post(url:String,body:String):String{val c=URL(url).openConnection() as HttpURLConnection;c.requestMethod="POST";c.doOutput=true;c.setRequestProperty("Content-Type","application/json");c.setRequestProperty("X-Session-Token",token);c.setRequestProperty("Origin",API);c.outputStream.use{it.write(body.toByteArray())};val code=c.responseCode;val s=read(if(code>=400)c.errorStream else c.inputStream);if(code !in 200..299)throw IOException("HTTP $code $s");return s}
    private fun read(i:InputStream?):String=i?.bufferedReader()?.readText()?:("")
    private fun ui(s:String)=runOnUiThread{Toast.makeText(this,s,Toast.LENGTH_SHORT).show()}
    private fun pick(){startActivityForResult(Intent(Intent.ACTION_OPEN_DOCUMENT).apply{addCategory(Intent.CATEGORY_OPENABLE);type="application/zip"},PICK_PROJECT)}
    override fun onActivityResult(r:Int,c:Int,d:Intent?){super.onActivityResult(r,c,d);if(r==PICK_PROJECT&&c==RESULT_OK){selectedUri=d?.data;show("Build")}}

    private fun startBuild(){io.execute{try{if(token.isNullOrBlank())token=JSONObject(get("$API/api/session",null)).optString("token");val zip=prepare();val b="----WZH${System.currentTimeMillis()}";val c=URL("$API/api/build").openConnection() as HttpURLConnection;c.requestMethod="POST";c.doOutput=true;c.connectTimeout=60000;c.readTimeout=120000;c.setRequestProperty("X-Session-Token",token);c.setRequestProperty("Origin",API);c.setRequestProperty("Content-Type","multipart/form-data; boundary=$b");c.outputStream.use{out->Multipart(out,b).apply{file("projectZip",zip.name,"application/zip",zip);text("buildType","apk");text("platform","android");text("clientId",prefs.getString("clientId",UUID.randomUUID().toString())!!);text("signRelease","true");text("dex2c","false");text("cacheMode","fresh");text("splashStyle","off");text("splashSpeed","normal");text("permissions",JSONArray(selectedPermissions).toString());if(sourceMode=="url")text("remoteUrl",prefs.getString("remoteUrl","")!!);end()}};val code=c.responseCode;val s=read(if(code>=400)c.errorStream else c.inputStream);if(code !in 200..299)throw IOException("HTTP $code $s");jobId=JSONObject(s).optString("jobId");refreshLimit();runOnUiThread{show("Progress")};poll()}catch(e:Exception){ui("Build failed: ${e.message}")}}}
    private fun prepare():File{val f=File(cacheDir,"wzh-project.zip");if(sourceMode=="zip"&&selectedUri!=null){contentResolver.openInputStream(selectedUri!!).use{input->FileOutputStream(f).use{out->input!!.copyTo(out)}};return f};ZipOutputStream(FileOutputStream(f)).use{z->if(sourceMode=="url")add(z,"index.html","<script>location.replace(${JSONObject.quote(prefs.getString("remoteUrl","")!!)});</script>") else if(codeLang=="html")add(z,"index.html",htmlCode!!.text.toString()) else createNative(z)};return f}
    private fun createNative(z:ZipOutputStream){val p=derivedPackage();var src=nativeCode!!.text.toString();src=src.replace(Regex("(?s)^\\s*package\\s+[\\w.]+;?\\s*"),"");src="package $p${if(codeLang=="java")";" else ""}\n\n$src";add(z,"settings.gradle", "pluginManagement { repositories { google(); mavenCentral(); gradlePluginPortal() } }\ndependencyResolutionManagement { repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS); repositories { google(); mavenCentral() } }\nrootProject.name='GeneratedApp'\ninclude ':app'\n");add(z,"build.gradle",if(codeLang=="kotlin")"plugins { id 'com.android.application' version '8.7.2' apply false\n id 'org.jetbrains.kotlin.android' version '1.9.24' apply false }" else "plugins { id 'com.android.application' version '8.7.2' apply false }");add(z,"app/build.gradle","plugins { id 'com.android.application'${if(codeLang=="kotlin")"\n id 'org.jetbrains.kotlin.android'" else ""} }\nandroid { namespace '$p'; compileSdk 35; defaultConfig { applicationId '$p'; minSdk 23; targetSdk 35; versionCode 1; versionName '1.0' }; buildTypes { release { minifyEnabled false } }; compileOptions { sourceCompatibility JavaVersion.VERSION_17; targetCompatibility JavaVersion.VERSION_17 }${if(codeLang=="kotlin")"; kotlinOptions { jvmTarget='17' }" else ""} }");add(z,"app/src/main/AndroidManifest.xml","<manifest xmlns:android=\"http://schemas.android.com/apk/res/android\"><application android:theme=\"@style/AppTheme\" android:label=\"Generated App\"><activity android:name=\".MainActivity\" android:exported=\"true\"><intent-filter><action android:name=\"android.intent.action.MAIN\"/><category android:name=\"android.intent.category.LAUNCHER\"/></intent-filter></activity></application></manifest>");add(z,"app/src/main/res/values/styles.xml", "<resources><style name=\"AppTheme\" parent=\"android:style/Theme.Material.NoActionBar\"><item name=\"android:fontFamily\">sans</item><item name=\"android:colorAccent\">#63E6D5</item></style></resources>");add(z,"app/src/main/"+(if(codeLang=="kotlin")"kotlin" else "java")+"/"+p.replace('.','/')+"/MainActivity."+(if(codeLang=="kotlin")"kt" else "java"),src)}
    private fun add(z:ZipOutputStream,n:String,s:String){z.putNextEntry(ZipEntry(n));z.write(s.toByteArray(StandardCharsets.UTF_8));z.closeEntry()}
    private fun poll(){val id=jobId?:return;io.execute{try{val o=JSONObject(get("$API/api/status/$id",token));runOnUiThread{status?.text="Job: $id\n${o.optString("status")}\n${o.optString("message",o.optString("log",""))}"};val st=o.optString("status").lowercase();if(st.contains("done")||st.contains("success")||st.contains("failed")||st.contains("error")||st.contains("cancel"))return@execute;handler.postDelayed({poll()},2000)}catch(_:Exception){handler.postDelayed({poll()},4000)}}}
    private fun cancelBuild(){jobId?.let{id->io.execute{try{post("$API/api/cancel/$id","{}");ui("Build cancelled")}catch(e:Exception){ui("Cancel failed: ${e.message}")}}}}
    private fun loadHistory(view:TextView){io.execute{try{val cid=prefs.getString("clientId","")!!;val raw=get("$API/api/history?clientId="+URLEncoder.encode(cid,"UTF-8"),token);runOnUiThread{view.text=raw}}catch(e:Exception){runOnUiThread{view.text="Could not load history.\n${e.message}"}}}}
    private fun help(){AlertDialog.Builder(this).setTitle("WZH 2 APK").setMessage("Native Android client for the WZH build service. It uses native views and HTTP instead of a WebView. If your server enforces a browser-only Origin allowlist, that backend policy must explicitly permit the native client.").setPositiveButton("OK",null).show()}
    private class Multipart(private val out:OutputStream,private val boundary:String){fun text(n:String,v:String){out.write("--$boundary\r\nContent-Disposition: form-data; name=\"$n\"\r\n\r\n$v\r\n".toByteArray())};fun file(n:String,fn:String,type:String,f:File){out.write("--$boundary\r\nContent-Disposition: form-data; name=\"$n\"; filename=\"$fn\"\r\nContent-Type: $type\r\n\r\n".toByteArray());f.inputStream().use{it.copyTo(out)};out.write("\r\n".toByteArray())};fun end(){out.write("--$boundary--\r\n".toByteArray())}}
}
